ironfish/src/rpc/routes/wallet/multisig/dkg/round1.ts

```diff
@@ -0,0 +1,89 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
+import { multisig } from '@ironfish/rust-nodejs'
+import * as yup from 'yup'
+import { RPC_ERROR_CODES, RpcValidationError } from '../../../../adapters'
+import { ApiNamespace } from '../../../namespaces'
+import { routes } from '../../../router'
+import { AssertHasRpcContext } from '../../../rpcContext'
+
+export type DkgRound1Request = {
+  // TODO `secretName` or `multisigSecretName`?
+  name: string
+  minSigners: number
+  participants: Array<{ identity: string }>
+}
+
+export type DkgRound1Response = {
+  encryptedSecretPackage: string
+  publicPackage: string
+}
+
+export const DkgRound1RequestSchema: yup.ObjectSchema<DkgRound1Request> = yup
+  .object({
+    name: yup.string().defined(),
+    minSigners: yup.number().defined(),
+    participants: yup
+      .array()
+      .of(yup.object({ identity: yup.string().defined() }).defined())
+      .defined(),
+  })
+  .defined()
+
+export const DkgRound1ResponseSchema: yup.ObjectSchema<DkgRound1Response> = yup
+  .object({
+    encryptedSecretPackage: yup.string().defined(),
+    publicPackage: yup.string().defined(),
+  })
+  .defined()
+
+routes.register<typeof DkgRound1RequestSchema, DkgRound1Response>(
+  `${ApiNamespace.wallet}/multisig/dkg/round1`,
+  DkgRound1RequestSchema,
+  async (request, node): Promise<void> => {
+    AssertHasRpcContext(request, node, 'wallet')
+
+    const { name, minSigners, participants } = request.data
+
+    // TODO check if this secret is already being used for an account (there's
+    // nothing inherently wrong with using the same secret for multiple
+    // multisig groups, but I think our code assumes a 1:1 relationship between
+    // secret names and account names)
+    const multisigSecret = await node.wallet.walletDb.getMultisigSecretByName(name)
+
+    if (!multisigSecret) {
+      throw new RpcValidationError(
+        `Multisig secret with name '${name}' not found`,
+        400,
+        RPC_ERROR_CODES.MULTISIG_SECRET_NOT_FOUND,
+      )
+    }
+
+    const participantIdentities = participants.map((p) => p.identity)
+    const selfIdentity = new multisig.ParticipantSecret(multisigSecret.secret)
+      .toIdentity()
+      .serialize()
+      .toString('hex')
+
+    if (!participantIdentities.includes(selfIdentity)) {
+      throw new RpcValidationError(
+        `List of participant identities must include the identity for '${name}' (${selfIdentity})`,
+        400,
+        RPC_ERROR_CODES.VALIDATION,
+      )
+    }
+
+    if (minSigners > participantIdentities.length) {
+      throw new RpcValidationError(
+        `minSigners (${minSigners}) exceeds the number of participants (${participantIdentities.length})`,
+        400,
+        RPC_ERROR_CODES.VALIDATION,
+      )
+    }
```

Comment:
Should we also verify that `minSigners >= 2`?

Reply:
Good idea, thanks for spotting that